File: lab4/Entities/Project.cs
using lab4.Helpers;

namespace lab4.Entities
{
    public class ProjectJson : EntityJson
    {
        public string projectId { get; set; } = "";
        public string name { get; set; } = "";
        public int budget { get; set; } = 0;
        public string managerName { get; set; } = "";
        public bool active { get; set; } = true;
        public ICollection<SubactivityJson>? subactivities { get; set; }
    }

    public class Project : IEntity<ProjectJson>
    {
        // PK
        public string ProjectId { get; set; } = "";

        public string Name { get; set; } = "";
        public int Budget { get; set; } = 0;
        public string ManagerName { get; set; } = "";
        public bool Active { get; set; } = true;

        // Parents
        public User? Manager { get; set; }

        // Children
        public virtual ICollection<ApprovedActivity>? ApprovedActivities { get; set; }
        public virtual ICollection<Subactivity>? Subactivities { get; set; }
        public virtual ICollection<UserActivity>? UserActivities { get; set; }

        public ProjectJson toJSON()
        {
            return new ProjectJson
            {
                projectId=ProjectId,
                name=Name,
                budget=Budget,
                managerName=ManagerName,
                active=Active,
                subactivities= Helper.toJSON<Subactivity, SubactivityJson>(this.Subactivities)
            };
        }
    }
}

File: lab4/ClientApp/src/entities/Project.ts
import Subactivity from "./Subactivity";

export class Project {
    projectId: string;
    name: string;
    budget: number;
    managerName: string;
    active: boolean;
    subactivities: Subactivity[];

    constructor(projectId: string, name: string, budget: number, managerName: string, active: boolean, subactivities: Subactivity[]) {
        this.projectId = projectId;
        this.name = name;
        this.budget = budget;
        this.managerName = managerName;
        this.active = active;
        this.subactivities = subactivities;
    }

    toJSON() {
        return {
            projectId: this.projectId,
            name: this.name,
            budget: this.budget,
            managerName: this.managerName,
            active: this.active,
            subactivities: this.subactivities,
        };
    }

    static fromJSON(json: any) {
        return new Project(json['projectId'], json['name'], json['budget'], json['managerName'], json['active'], json['subactivities']);
    }

    static createEmpty() {
        return new Project('', '', 0, '', true, []);
    }
};

export default Project;
